File: platform/lib-node/src/x-canary-consume-context.ts
import { runWithCanary } from "./x-canary-context.js";
import { isCanaryFlagged, type KafkaConsumeHeaders } from "./x-canary-consume-filter.js";

export async function runWithCanaryFromHeaders<T>(
  headers: KafkaConsumeHeaders,
  handler: () => Promise<T>,
): Promise<T> {
  const canary = isCanaryFlagged(headers);
  return runWithCanary(canary, handler);
}

File: platform/lib-java/src/main/java/com/canary/platform/lib/XCanaryConsumeContext.java
package com.canary.platform.lib;

import org.apache.kafka.common.header.Headers;

/**
 * Helper for Kafka consume callbacks: opens an XCanaryContext frame
 * around the handler so outbound HTTP/Kafka/Restate calls inherit
 * x-canary via the existing Phase 1 interceptors.
 */
public final class XCanaryConsumeContext {

    private XCanaryConsumeContext() {}

    public static void runWith(Headers kafkaHeaders, Runnable handler) {
        boolean canary = XCanaryConsumeFilter.isCanaryFlagged(kafkaHeaders);
        boolean prior = XCanaryContext.isCanary();
        XCanaryContext.set(canary);
        try {
            handler.run();
        } finally {
            XCanaryContext.set(prior);
            if (!prior) XCanaryContext.clear();
        }
    }
}
